File: apps/demo-be-csas/src/DemoBeCsas/Infrastructure/AppDbContext.cs
using DemoBeCsas.Domain;
using DemoBeCsas.Infrastructure.Models;
using Microsoft.EntityFrameworkCore;

namespace DemoBeCsas.Infrastructure;

public class AppDbContext(DbContextOptions<AppDbContext> options) : DbContext(options)
{
    public DbSet<UserModel> Users => Set<UserModel>();
    public DbSet<ExpenseModel> Expenses => Set<ExpenseModel>();
    public DbSet<AttachmentModel> Attachments => Set<AttachmentModel>();
    public DbSet<RevokedTokenModel> RevokedTokens => Set<RevokedTokenModel>();

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<UserModel>(entity =>
        {
            entity.HasKey(e => e.Id);
            entity.HasIndex(e => e.Username).IsUnique();
            entity.HasIndex(e => e.Email).IsUnique();
            entity.Property(e => e.Status).HasConversion<string>();
            entity.Property(e => e.Role).HasConversion<string>();
        });

        modelBuilder.Entity<ExpenseModel>(entity =>
        {
            entity.HasKey(e => e.Id);
            entity.Property(e => e.Amount).HasPrecision(18, 6);
            entity.Property(e => e.Type).HasConversion<string>();
        });

        modelBuilder.Entity<AttachmentModel>(entity =>
        {
            entity.HasKey(e => e.Id);
        });

        modelBuilder.Entity<RevokedTokenModel>(entity =>
        {
            entity.HasKey(e => e.Id);
            entity.HasIndex(e => e.Jti).IsUnique();
            entity.HasIndex(e => e.UserId);
        });
    }
}

File: apps/organiclever-be/src/OrganicLeverBe/Infrastructure/AppDbContext.fs
module OrganicLeverBe.Infrastructure.AppDbContext

open System
open System.ComponentModel.DataAnnotations
open System.ComponentModel.DataAnnotations.Schema
open Microsoft.EntityFrameworkCore

[<CLIMutable>]
[<Table("users")>]
type UserEntity =
    { [<Key>]
      [<Column("id")>]
      Id: Guid
      [<Column("email")>]
      Email: string
      [<Column("name")>]
      Name: string
      [<Column("avatar_url")>]
      AvatarUrl: string
      [<Column("google_id")>]
      GoogleId: string
      [<Column("created_at")>]
      CreatedAt: DateTime
      [<Column("updated_at")>]
      UpdatedAt: DateTime }

[<CLIMutable>]
[<Table("refresh_tokens")>]
type RefreshTokenEntity =
    { [<Key>]
      [<Column("id")>]
      Id: Guid
      [<Column("user_id")>]
      UserId: Guid
      [<Column("token_hash")>]
      TokenHash: string
      [<Column("expires_at")>]
      ExpiresAt: DateTime
      [<Column("created_at")>]
      CreatedAt: DateTime }

type AppDbContext(options: DbContextOptions<AppDbContext>) =
    inherit DbContext(options)

    [<DefaultValue(false)>]
    val mutable users: DbSet<UserEntity>

    member this.Users
        with get () = this.users
        and set v = this.users <- v

    [<DefaultValue(false)>]
    val mutable refreshTokens: DbSet<RefreshTokenEntity>

    member this.RefreshTokens
        with get () = this.refreshTokens
        and set v = this.refreshTokens <- v

    override this.OnModelCreating(modelBuilder: ModelBuilder) =
        modelBuilder.Entity<UserEntity>().HasIndex(fun u -> u.Email :> obj).IsUnique()
        |> ignore

        modelBuilder.Entity<UserEntity>().HasIndex(fun u -> u.GoogleId :> obj).IsUnique()
        |> ignore

        modelBuilder.Entity<RefreshTokenEntity>().HasIndex(fun t -> t.TokenHash :> obj).IsUnique()
        |> ignore

        modelBuilder.Entity<UserEntity>().Property(fun u -> u.AvatarUrl).IsRequired(false)
        |> ignore

        base.OnModelCreating(modelBuilder)
